### DISCUSSION TITLE
Date Picker acting strange

### DISCUSSION THREAD

**Discussion by Rodien** [2sic/2sxc#2441] [Q&A] [answered]
Hi guys,

I am having an issue with date picker not showing the right local time. 

The date picker time is 2 hours behind.

I tried using formulas on a normal text field and the date is showing correctly. 
So, the issue seems to be with the date picker...?

`v1(data) {
let current_datetime = new Date()
let formatted_date = current_datetime.getFullYear() + "-" + (current_datetime.getMonth() + 1) + "-" + current_datetime.getDate() + " " + current_datetime.getHours() + ":" + current_datetime.getMinutes() + ":" + current_datetime.getSeconds() 

  return data.prefill =formatted_date;
}`

![image](https://user-images.githubusercontent.com/50739829/120973397-291d1080-c76f-11eb-92e8-9e3131b19e8e.png)

Is this a known issue with the date picker?

**Comment by iJungleboy** [comment]
@Rodien Thanks - and awesome that you're already experimenting with Formulas.

We actually haven't fully tested the date/time formulas yet, so this may be an issue. I believe our goal is make a change that dates will be as date-time object (not as strings) whichs should then fix this. But it probably means that the input/output format may still need to change in the date-formulas. 

For now, could you check if it would be fixed if you return the date-string as Zulu time?

**Reply by Rodien** [reply]
@iJungleboy Formulas are really great❗



![image](https://user-images.githubusercontent.com/50739829/121003732-27167a00-c78e-11eb-8cdb-5e952c080865.png)

Both fields are using the same formula. 
returnug string to the date field does not work.

```
v1(data) {
  let d = new Date();
  let data.prefill= d.toLocaleString('nl-NL', { timeZone: 'Zulu' });
}
```

**Comment by PagnoDunadan** [comment]
Date picker uses UTC time as a string and you are giving it your local time. So you local time is converted to UTC. This code will give you correct result until we work out DateTime as an object:

```
v1 (data) {
  const date = new Date();
  const result = date.toJSON();
  return result;
}
```

**Reply by Rodien** [reply]
returning date.toJSON() is still giving me localtime -2 

![image](https://user-images.githubusercontent.com/50739829/121005384-0cdd9b80-c790-11eb-8874-c1145059ac67.png)


I will try some other things in the meantime.

**Comment by iJungleboy** [comment]
We'll try to address this asap. 
Note that formulas in 12.02 about to be released should then use a date-object and not a date string (WIP).

**Reply by Rodien** [reply]
@iJungleboy THANKS

I just want to point out that I don't think that the formulas are at fault here. 

The default behavior of the datetime picker is not taking DNN current culture into consideration.

Just like @PagnoDunadan mention datetime picker is returning UTC by default. 
It should translate it base on DNN current Timezone settings; I think?

**Comment by iJungleboy** [comment]
Just fyi: in the end it must be the right time on UTC, not on the local PC. 

You can fiddle with TimeZoneOffset to fix that or something like this:

```
function createDateAsUTC(date) {
    return new Date(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate(), date.getHours(), date.getMinutes(), date.getSeconds()));
}
```

**Reply by Rodien** [reply]
Adding ".toJSON()" at the end did it for me. 👏

Without ".toJSON()" it was showing current time +2 hours. 


Thanks guys @iJungleboy  @PagnoDunadan 💖

```
 let newDate = new Date(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate(), date.getHours(), date.getMinutes(), date.getSeconds())).toJSON();


  return data.prefill = newDate ;
```

**Reply by iJungleboy** [reply]
Note that your return-statement looks strange. It should either just return the newDate or you would need a == instead of an =

**Comment by iJungleboy** [comment, marked as answer]
Note in v12.02 we are making date-formulas easier. 

1. If your formula returns a proper date object, we assume you didn't think about time zones and we simply drop them. So if you return a date at 14:00 we assume it's supposed to be 14:00 even if it says GMT+0200 or something (that is removed)
2. If your formula returns Zulu time that's the time it will use
3. If your formula returns a string with a time-zone shift, that too will be ignored/removed

See latest docs http://r.2sxc.org/formulas

I also documented in detail how the dates behave, and how to parse an existing date (dropping the Z) to ensure it's easy to work with.

**Reply by Rodien** [reply]
GREAT WORK 🎉😁👏
